### PR TITLE
Initial commit of Rebar3 Nix Bootstrap

### PR DESCRIPTION
**Pull request by ericbmerritt**
This is a very simply 'shim' project that exists to help rebar3 work
in the context of Nix. Its purpose is to bootstrap a project so that
rebar3 doesn't try to manage dependencies (that should be left to
Nix).

This does the following work.
- It copies the hex registry to a place where rebar3 can find it
- It copies any dependencies into the `_build` dir to a place where
  rebar3 can find it.
- It updates the `*.app.src` with the version that is available to
  from the nix expression.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,22 @@
...
+
+This is a very simply 'shim' project that exists to help rebar3 work
+in the context of Nix. Its purpose is to bootstrap a project so that
+rebar3 doesn't try to manage dependencies (that should be left to
+Nix).
+
+This does the following work.
+
+# It copies the hex registry to a place where rebar3 can find it
+# It copies any dependencies into the `_build` dir to a place where
```
Comment (gleber) [inline README.md]: Is it possible to make it link things, not copy? Ideally there is no need to create copies and reuse files from /nix/store/

File: README.md [hunk 2]
```diff
@@ -1,2 +1,22 @@
...
+  rebar3 can find it.
+# It updates the `*.app.src` with the version that is available to
```
Comment (gleber) [inline README.md]: Why is this necessary? Shouldn't content of .app.src be in sync with nix expression?

File: README.md [hunk 3]
```diff
@@ -1,2 +1,22 @@
...
+  from the nix expression.
+
+Usage
+----
+
+Simply run `rebar3-nix-bootstrap` inside of a `nix-shell` with
+relevant Erlang support. It will figure things out and do the right
```
Comment (gleber) [inline README.md]: Can you either provide a shell command to enter such a shell or add a link to other docs which contains the command?

File: src/rebar3-nix-bootstrap [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+        true ->
+            ok;
+        false ->
+            stderr("Registry snapshot (~s) does not exist!", [RegistrySnapshot])
+    end.
+
+-spec gather_required_data_from_the_environment(#data{}) -> {ok, map()}.
+gather_required_data_from_the_environment(ArgData) ->
+    {ok, ArgData#data{ version = guard_env("version")
+                     , erl_libs = os:getenv("ERL_LIBS", [])
```
Comment (gleber) [inline src/rebar3-nix-bootstrap:143]: Why isn't this also guard_env?

File: src/rebar3-nix-bootstrap [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+    ok = file:make_symlink(RegistrySnapshot,
+                           ?HEX_REGISTRY_PATH).
+
+-spec make_sure_registry_snapshot_exists(string()) -> ok.
+make_sure_registry_snapshot_exists(RegistrySnapshot) ->
+    case filelib:is_file(RegistrySnapshot) of
+        true ->
+            ok;
+        false ->
+            stderr("Registry snapshot (~s) does not exist!", [RegistrySnapshot])
```
Comment (gleber) [inline src/rebar3-nix-bootstrap]: Shouldn't it exit with non-zero error code here (or inside stderr)?

**Comment by gleber** [comment]
I've added some comments, nothing critical. 

I hope we can eventually move this code into rebar itself or into a rebar plugin (which would somehow automagically hook into correct spots and do all this for the user). This is a great start, though!

**Comment by ericbmerritt** [comment]
Hopefully, I have had pretty extensive conversations with Tristan and they are really not interested in this functionality at all. :/ hopefully, this gets popular and that changes though.

File: README.md (hunk 1 above)
Comment (ericbmerritt) [inline README.md]: It actually does link not copy. I will update the readme so its accurate.

File: README.md (hunk 2 above)
Comment (ericbmerritt) [inline README.md]: It does that because thats what you where doing in the original build-hex.nix function. I also think its a good idea. Simply because the app should reflect the nix version. Now the question, I guess could be which version should drive the other.

File: README.md (hunk 3 above)
Comment (ericbmerritt) [inline README.md]: Are you talking about `nix-shell`. If so thats a pretty fundamental part of Nix tooling. I suspect folks looking at this would be familiar with it. That said, I will provide a link.

File: src/rebar3-nix-bootstrap (hunk 4 above)
Comment (ericbmerritt) [inline src/rebar3-nix-bootstrap:143]: Because if it doesn't exist, thats ok it just doesn't do any app linking. It will still do the *.app.src part of the bootstrap.

File: src/rebar3-nix-bootstrap (hunk 5 above)
Comment (ericbmerritt) [inline src/rebar3-nix-bootstrap]: good, catch. Yes it should.

File: src/rebar3-nix-bootstrap (hunk 4 above)
Comment (gleber) [inline src/rebar3-nix-bootstrap:143]: Based on https://github.com/NixOS/nixpkgs/compare/master...ericbmerritt:erlang-improvements#diff-85631298b9cdea5c4f43bb9be3fd5a93R37 it seem that ERL_LIBS should always be defined if the tool is called correctly, even if there are no deps. I prefer stricter code when possible, since it'll fail faster, but it's up to you.

File: README.md (hunk 2 above)
Comment (gleber) [inline README.md]: Oh, indeed. I've actually taken this code from https://github.com/sjmackenzie/nix-ejabberd/blob/master/build-support/build-erlang.nix#L18 and forgot about it.

I am not sure which one should drive the other actually... Maybe instead add an assert if the are out of sync?

**Comment by ericbmerritt** [comment]
Updated the things that made sense to update.

File: README.md (hunk 3 above)
Comment (gleber) [inline README.md]: Oh, sorry I was not clear enough. I meant adding an example like

```
nix-shell -A erlangPackages.ibrowse.env --run "rebar3-nix-bootstrap"
```

Or even better, once we have rebar2nix command, something like:

```
rebar2nix --shell > shell.nix
nix-shell --run "rebar3-nix-bootstrap"
```

WDYT?

File: src/rebar3-nix-bootstrap [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+    ok = bootstrap_libs(RequiredData).
+
+%% @doc
+%% Argument parsing is super simple only because we want to keep the
+%% dependencies minimal. For now there can be two entries on the
+%% command line, "register-only" and "compile-ports"
+-spec parse_args([string()]) -> #data{}.
+parse_args(Args) ->
+    {ok, #data{registry_only = lists:member("registry-only", Args)
+              , compile_ports = lists:member("compile-ports", Args)}}.
```
Comment (gleber) [inline src/rebar3-nix-bootstrap:60]: This does not seem to be mentioned at all from https://github.com/NixOS/nixpkgs/compare/master...ericbmerritt:erlang-improvements

Shouldn't it be called from there? Or maybe replace arg-checking with env-checking?

File: src/rebar3-nix-bootstrap (hunk 4 above)
Comment (ericbmerritt) [inline src/rebar3-nix-bootstrap:143]: I think ERL_LIBS is defined in Erlang now (its one of the changes that seems to be causing this command some problems) so I will do that. Just remember that the makefile for this project isn't run when the code above is run. The code above is run during the build of packages that use it.

File: src/rebar3-nix-bootstrap (hunk 6 above)
Comment (ericbmerritt) [inline src/rebar3-nix-bootstrap:60]: yup . fixed.
